src/Reassem.h

```diff
@@ -20,30 +22,235 @@ enum ReassemblerType {
 
 class Reassembler;
 
+
+/**
+ * A block/segment of data for use in the reassembly process.
+ */
 class DataBlock {
 public:
-	DataBlock(Reassembler* reass, const u_char* data,
-	          uint64_t size, uint64_t seq,
-	          DataBlock* prev, DataBlock* next,
-	          ReassemblerType reassem_type = REASSEM_UNKNOWN);
 
-	~DataBlock();
+	/**
+	 * Create a data block/segment with associated sequence numbering.
+	 */
+	DataBlock(const u_char* data, uint64_t size, uint64_t seq);
+
+	DataBlock(const DataBlock& other)
+		{
+		seq = other.seq;
+		upper = other.upper;
+		auto size = other.Size();
+		block = new u_char[size];
+		memcpy(block, other.block, size);
+		}
+
+	DataBlock(DataBlock&& other)
+		{
+		seq = other.seq;
+		upper = other.upper;
+		block = other.block;
+		other.block = nullptr;
+		}
+
+	DataBlock& operator=(const DataBlock& other)
+		{
+		if ( this == &other )
+			return *this;
+
+		seq = other.seq;
+		upper = other.upper;
+		auto size = other.Size();
+		delete [] block;
+		block = new u_char[size];
+		memcpy(block, other.block, size);
+		return *this;
+		}
+
+	DataBlock& operator=(DataBlock&& other)
+		{
+		if ( this == &other )
+			return *this;
+
+		seq = other.seq;
+		upper = other.upper;
+		delete [] block;
+		block = other.block;
+		other.block = nullptr;
+		return *this;
+		}
+
+	~DataBlock()
+		{ delete [] block; }
+
+	/**
+	 * @return length of the data block
+	 */
+	uint64_t Size() const
+		{ return upper - seq; }
+
+	uint64_t seq;
+	uint64_t upper;
+	u_char* block;
+};
 
-	uint64_t Size() const	{ return upper - seq; }
+using DataBlockMap = std::map<uint64_t, DataBlock>;
 
-	DataBlock* next;	// next block with higher seq #
-	DataBlock* prev;	// previous block with lower seq #
-	uint64_t seq, upper;
-	u_char* block;
-	ReassemblerType rtype;
 
-	Reassembler* reassembler; // Non-owning pointer back to parent.
+/**
+ * The data structure used for reassembling arbitrary sequences of data
+ * blocks/segments.  It internally uses an ordered map (std::map).
+ */
+class DataBlockList {
+public:
+
+	DataBlockList()
+		{ }
+
+	DataBlockList(Reassembler* r) : reassembler(r)
+		{ }
+
+	~DataBlockList()
+		{ Clear(); }
+
+	/**
+	 * @return iterator to start of the block list.
+	 */
+	DataBlockMap::const_iterator Begin() const
+		{ return block_map.begin(); }
+
+	/**
+	 * @return iterator to end of the block list (one past last element).
+	 */
+	DataBlockMap::const_iterator End() const
+		{ return block_map.end(); }
+
+	/**
+	 * @return reference to the first data block in the list.
+	 * Must not be called when the list is empty.
+	 */
+	const DataBlock& FirstBlock() const
+		{ return block_map.begin()->second; }
+
+	/**
+	 * @return reference to the last data block in the list.
+	 * Must not be called when the list is empty.
+	 */
+	const DataBlock& LastBlock() const
```

Comment:
The cumbersome part is mostly the repetition of `std::prev(block_list.End())->second`, which I replaced with `block_list.LastBlock()`, but I prefer the explicit logic to check for "empty list" rather than indirectly via "null last block".  Not really opposed to adding a new method for that, but just going to leave it up to whoever else next works on the code to add if it's useful to them.